### ISSUE TITLE
When compiling Marlin RC7 or the newest RCBugFix with MAX31855, Marlin goes into some kind of loop and doesn't respond any more

### ISSUE THREAD

**Issue by derletztename** [MarlinFirmware/Marlin#4938]
When I try to use Marlin on my RAMPS 1.4 EFB with the MAX31855 thermocouple board and the thermocouple after compiling and uploading, Marlin goes into some kind of loop and I can't connect with pronterface. Also I can't upload anything again unless I upload just in the tiny moment after plugging the Arduino into my PC. This happens with Marlin RC7 and RCBugFix on every Arduino I tested with and without RAMPS. And it only happens when I compile Marlin with the "-3" for MAX31855 support.
Maybe someone could figure out, what is wrong?
I'm printing metal, I need to measure high temperatures :(

```
//// Temperature sensor settings:
// -3 is thermocouple with MAX31855 (only for sensor 0)
```

**Comment by thinkyhead** [comment]
The most likely thing getting stuck related to that thermocouple is the reading of the `MAX31855` itself. The loop which reads the bits will stick if it never gets all 4 bytes. So perhaps it's an issue of timing. The read speed for the MAX31855 is 4x the rate of the MAX6675, so possibly the initial delay is too large. Try editing the method `Temperature::read_max6675` in `temperature.cpp` and removing one of these two "nop" lines…

``` cpp
    // ensure 100ns delay - a bit extra is fine
    asm("nop");//50ns on 20Mhz, 62.5ns on 16Mhz
    asm("nop");//50ns on 20Mhz, 62.5ns on 16Mhz
```

**Comment by derletztename** [comment]
I am not a coder and I have no idea what it really means that you just told me, but I will try this out and post my experiences. Thank you :)

**Comment by thinkyhead** [comment]
I hear ya – I've never been to Moscow. Let me know if the suggestion helps.

**Comment by derletztename** [comment]
It did not change anything at all, as it seem :(

Actually both ICs look very similar:
https://cdn-shop.adafruit.com/datasheets/MAX31855.pdf
https://cdn-shop.adafruit.com/datasheets/MAX6675.pdf

The only easy to see difference is:
MAX6675
The data is output in a 12-bit resolution, SPI™-compatible, read-only format.
MAX31855
The  data  is  output  in  a  signed  14-bit,  SPI-compatible,  read-only  format. 

Also the MAX6675 can only work with type-K thermocouples and the MAX31855 works with several thermocouple types, which is why I am trying to use it...

Is this a problem for Marlin somehow, that it treats both ICs the same? 12-bit resolution and signed  14-bit?

**Comment by derletztename** [comment]
Here are two Arduino Librarys for the MAX6675 and the MAX31855. Maybe the difference in the code helps fixing the Marlin Issues?
https://learn.adafruit.com/thermocouple/using-a-thermocouple

**Comment by derletztename** [comment]
The MAX31855 code is twice the lines than the MAX6675. Also it looks different. Nevertheless it's still russian to me :(

**Comment by Grogyan** [comment]
In temperature,cpp line 1293 
#define MAX6675_SPEED_BITS (_BV(SPR1)) // clock ÷ 64
You can try changing SPR1 to SPR0.  This increases the clock rate for the MAX31855, but you may end up with erranous values at times.

Can you please post your configuration.h file.

Also if you are in a particuarly noisy location, you may have spikes on the data/clock lines or even the thermocouple itself.  Also note that that you need to plug in the 5V from RAMPS to the 5V pin on the MAX31855 board, not the 3V3 rail, as that is an output .

**Comment by derletztename** [comment]
[config_h.txt](https://github.com/MarlinFirmware/Marlin/files/514718/config_h.txt)

**Comment by derletztename** [comment]
I am not using the IC directly, I am using the Adafruit breakout board
https://www.adafruit.com/index.php?main_page=product_info&products_id=269&zenid=516505f41182f3a344fea2f8955968e0

**Comment by Grogyan** [comment]
This line should be uncommented

``` cpp
//#define THERMAL_PROTECTION_HOTENDS // Enable thermal protection for all extruders
```

This line should be a lower value

``` cpp
#define HEATER_0_MAXTEMP 1650
```

K type typically has a 500 degree upper limit, what thermocouple are you actually using?

**Comment by derletztename** [comment]
I'm printing Steel, it's very experimental ;)
For the first experiments I only used a small amount of power and therefore it's heating up very slow. That's a problem for the thermal runaway...

**Comment by derletztename** [comment]
The System can melt steel easily, but I wasn't able to measure higher temperatures above 1350°C, as I can't get that stupid MAX31855 to work :(

**Comment by Grogyan** [comment]
@derletztename what thermocouple are you using?

This from the adafruit site
Works with any K type thermocouple
Will not work with any other kind of thermocouple other than K type

**Comment by derletztename** [comment]
started with a thermocouple type k, and now i am trying to use a thermocouple type s, which only works with the MAX31855.
I have the breakout board with a thermocouple type k for testing, but also changed the IC on another board and try to use it with a chinese thermocouple type S...

**Comment by derletztename** [comment]
Here is a picture of the system melting steel wire:
![wp_20160610_22_34_49_rich](https://cloud.githubusercontent.com/assets/7712645/19168784/12f154b2-8c12-11e6-8e42-3abdd3c5f276.jpg)

**Comment by Grogyan** [comment]
https://datasheets.maximintegrated.com/en/ds/MAX31855.pdf

package for S type thermocouple should be
MAX31855SASA+T

**Comment by derletztename** [comment]
I asked the adafruit technicans. They say if you change the chip (solder, unsolder) it's suppose to work. Nevertheless it's not even working with the type k thermocouple without changing anything. I bought 2 breakout boards to tests this

**Comment by derletztename** [comment]
I postet the same pdf^^

**Comment by Grogyan** [comment]
The max6675 also has an alternate type of package

on the SO8 it should read MAX31855S  whereas the original chip marking reads MAX31855K
The K and S are the cold junction offsets within the device

**Comment by Grogyan** [comment]
Can you also please post a photo of your thermocouple boards?

**Comment by derletztename** [comment]
https://cdn-shop.adafruit.com/datasheets/MAX6675.pdf says it's only reading type K thermocouples. Or how do you mean alternate type of package?

**Comment by derletztename** [comment]
I use the adafruit board.
https://www.adafruit.com/index.php?main_page=product_info&products_id=269&zenid=516505f41182f3a344fea2f8955968e0 and plugged it in as said on their site

**Comment by Grogyan** [comment]
My mistake, I thought there was another variant of the 6675 for different thermocouples

**Comment by derletztename** [comment]
Man, you brought up hopes :(
Is there hope to get the MAX31855 running?

**Comment by Grogyan** [comment]
Indeed, I suggest first using an unmodified MAX31855 board and use a fresh copy of the Marlin firmware

**Comment by derletztename** [comment]
I did. I bought 2 breakout boards and only modified one. I am trying to get it working with the unmodified board and I hardly changed anything in the Marlin RC and RCBugFix edition.
The error also happens without the RAMPS plugged in the Arduino. But the error do not happen without the -3 for the MAX31855 board set. So it must be an error in the Marlin code I guess

**Comment by derletztename** [comment]
Just try this, download a new copy of Marlin and compile it with the -3 for the MAX31855 and see if you can connect to your Arduino afterwards. For me it don't work with pronterface nor with the Arduino software itself anymore. The only workaround is to upload another code just after plugging the Arduino in

**Comment by Grogyan** [comment]
I can't try that out till late tonight when I get home from work

**Comment by Blue-Marlin** [comment]
Might not bee very interesting for your problem.
To make negative temperature readings correct with a MAX31855, these chanes are needed.
Shifting right a negative unsigned will not shift in the needed 1s but always 0s (what makes them high positives).

``` diff
@@ -1295,16 +1295,16 @@ void Temperature::disable_all_heaters() {
 #if ENABLED(HEATER_0_USES_MAX6675)

   #define MAX6675_HEAT_INTERVAL 250u

   #if ENABLED(MAX6675_IS_MAX31855)
-    uint32_t max6675_temp = 2000;
+    int32_t max6675_temp = 2000;
     #define MAX6675_ERROR_MASK 7
     #define MAX6675_DISCARD_BITS 18
     #define MAX6675_SPEED_BITS (_BV(SPR1)) // clock ÷ 64
   #else
-    uint16_t max6675_temp = 2000;
+    int16_t max6675_temp = 2000;
     #define MAX6675_ERROR_MASK 4
     #define MAX6675_DISCARD_BITS 3
     #define MAX6675_SPEED_BITS (_BV(SPR0)) // clock ÷ 16
   #endif

@@ -1347,11 +1347,11 @@ void Temperature::disable_all_heaters() {
     if (max6675_temp & MAX6675_ERROR_MASK)
       max6675_temp = 4000; // thermocouple open
     else
       max6675_temp >>= MAX6675_DISCARD_BITS;

-    return (int)max6675_temp;
+    return max6675_temp;
   }

 #endif //HEATER_0_USES_MAX6675
```

**Comment by Blue-Marlin** [comment]
For analysing the problem

``` diff
@@ -1343,11 +1343,31 @@ void Temperature::disable_all_heaters() {
     }

     WRITE(MAX6675_SS, 1); // disable TT_MAX6675

     if (max6675_temp & MAX6675_ERROR_MASK)
-      max6675_temp = 4000; // thermocouple open
+     #if ENABLED(MAX6675_IS_MAX31855)
+     if (max6675_temp & 1) {
+       SERIAL_ERROR_START;
+        SERIAL_ERRORLNPGM("Temp measurement error! MAX31855 Open Circuit");
+     }
+     if (max6675_temp & 2) {
+       SERIAL_ERROR_START;
+        SERIAL_ERRORLNPGM("Temp measurement error! MAX31855 Short to GND");
+     }
+     if (max6675_temp & 4) {
+       SERIAL_ERROR_START;
+        SERIAL_ERRORLNPGM("Temp measurement error! MAX31855 Short to VCC");
+     }
+     #else
+     if (max6675_temp & 4) {
+       SERIAL_ERROR_START;
+        SERIAL_ERRORLNPGM("Temp measurement error! MAX6675");
+     }
+     #endif
+      //max6675_temp = 4000; // thermocouple open
+      max6675_temp = 25; // JUST TO AVOID THE KILL!!! Don't use in production.
     else
       max6675_temp >>= MAX6675_DISCARD_BITS;

     return (int)max6675_temp;
   }
```

might help.

Wo bist den Du in Deutschland?

**Comment by derletztename** [comment]
Nähe Köln/Bonn. Über Support vor Ort würde ich mich seeehr freuen ;)

**Comment by derletztename** [comment]
> #endif
> 
> ``` diff
> +      //max6675_temp = 4000; // thermocouple open
> +      max6675_temp = 25; // JUST TO AVOID THE KILL!!! Don't use in production.
>      else
>        max6675_temp >>= MAX6675_DISCARD_BITS;
> ```
> 
> --> leads to an error because the "#endif" ends the "if loop" and than there is an else without an if. Is the position, how much the line is indented (eingerückt) important for Arduino Code? Or should this be solved by the "#"? I hate to have no idea about coding :(

**Comment by thinkyhead** [comment]
> leads to an error because the "#endif" ends the "if loop"

Try this:

``` cpp
    if (max6675_temp & MAX6675_ERROR_MASK) {
      SERIAL_ERROR_START;
      SERIAL_ERRORPGM("Temp measurement error! ");
      #if MAX6675_ERROR_MASK == 7
        SERIAL_ERRORPGM("MAX31855 ");
        if (max6675_temp & 1)
          SERIAL_ERRORLNPGM("Open Circuit");
        else if (max6675_temp & 2)
          SERIAL_ERRORLNPGM("Short to GND");
        else if (max6675_temp & 4)
          SERIAL_ERRORLNPGM("Short to VCC");
      #else
        SERIAL_ERRORLNPGM("MAX6675");
      #endif
      //max6675_temp = 4000; // thermocouple open
      max6675_temp = 25; // JUST TO AVOID THE KILL!!! Don't use in production.
    }
    else
      max6675_temp >>= MAX6675_DISCARD_BITS;
```

**Comment by derletztename** [comment]
At least this is compiling, going to test this on the machine soon :)

**Comment by ghost** [comment]
I've completely same trouble as @derletztename.
Arduino MEGA2560 + RAMPS 1.4 EFB + [Adafruit MAX31855](https://www.adafruit.com/products/269).
I doubted that my hardware is broken at [2 month ago](https://github.com/MarlinFirmware/Marlin/issues/4533#issuecomment-239603375), but now I setup new MEGA2560 + new RAMPS, same trouble still alive.

It continuously reboots (freeze) when using MAX31855 without `SDSUPPORT`.
When freezing, red LED on RAMPS is flashing.
It can be reproduced without real thing of MAX31855. 

When I enable `SDSUPPORT`, MINTENP Err is occurs on every startup.

> Try this:

Unfortunately any message of those was shown.

**Comment by ghost** [comment]
> ``` diff
>     // ensure 100ns delay - a bit extra is fine
> -    asm("nop");//50ns on 20Mhz, 62.5ns on 16Mhz
> -    asm("nop");//50ns on 20Mhz, 62.5ns on 16Mhz
> +    //asm("nop");//50ns on 20Mhz, 62.5ns on 16Mhz
> +    //asm("nop");//50ns on 20Mhz, 62.5ns on 16Mhz
> ```
> 
> ``` diff
> -    uint32_t max6675_temp = 2000;
> +    int32_t max6675_temp = 2000;
> ```
> 
> ``` diff
> -    uint16_t max6675_temp = 2000;
> +    int16_t max6675_temp = 2000;
> ```
> 
> ``` diff
> -    return (int)max6675_temp;
> +    return max6675_temp;
> ```

These weren't also solved problem unfortunately.

**Comment by ghost** [comment]
~~And, I'm seeing interesting result.~~
~~With [this setting](https://github.com/esenapaj/Marlin/commit/5255213aea4a9e7b1c9f1a0d143ca71e51c07cbb), MINTENP Err is occurs on startup.~~
~~But with [this setting](https://github.com/esenapaj/Marlin/commit/334f39c7844e088ebc5b1287d609d232ec342ceb), MAXTEMP Err and MAX31855 Open Circuit Err are occurs on startup.~~

~~It seems that something of other setting is affecting to result.~~

Edit: My mistake of test. Results are same.
Above Open Circuit Err occurred when MAX31855 was not connected.

**Comment by ghost** [comment]
Still searching, but [this setting and dirty hack](https://github.com/esenapaj/Marlin/commit/a121cbfa5dc0164955b617e99291cb4e409cdeae) works fine for some reason.

``` diff
  #if ENABLED(HEATER_0_USES_MAX6675)

-    #if DISABLED(SDSUPPORT)
+    //#if DISABLED(SDSUPPORT)
      OUT_WRITE(SCK_PIN, LOW);
      OUT_WRITE(MOSI_PIN, HIGH);
      SET_INPUT(MISO_PIN);
-     WRITE(MISO_PIN,1);
-    #else
+    WRITE(MISO_PIN, HIGH);
+    //#else
      OUT_WRITE(SS_PIN, HIGH);
-    #endif
+    //#endif

    OUT_WRITE(MAX6675_SS, HIGH);

  #endif //HEATER_0_USES_MAX6675
```

**Comment by Grogyan** [comment]
I tried compiling RC7 a few days ago,  not RCBugFix, with Max31855
thermocouple and I just get a compiling error.
DIODIO_MISO had not defined in this scope. It was referring to pin 53 not
being defined,  possibly due to a conflict with SDCard,  it read like it
was a typo.
DIODIO_MISO should be DIO_MISO

I hadn't tried yet with RCBugFix,  I've not had a good week.

**Comment by ghost** [comment]
@Grogyan 

> DIODIO_MISO

That was a [bug in RC7](https://github.com/MarlinFirmware/Marlin/pull/4554), and fixed in current RCBugFix.

**Comment by ghost** [comment]
So, this hack (same as above) works fine with `SDSUPPORT` too, but there is a conditional.
When SDCard or microSD - SDCard adaptor is **already inserted in card slot at bootup** of Marlin, it causes freeze and LED flashing.
But SDCard or microSD - SDCard adaptor is **inserted to card slot after bootup** of Marlin, both SDCard and MAX31855 works fine.

``` diff
  #if ENABLED(HEATER_0_USES_MAX6675)

-    #if DISABLED(SDSUPPORT)
+    //#if DISABLED(SDSUPPORT)
      OUT_WRITE(SCK_PIN, LOW);
      OUT_WRITE(MOSI_PIN, HIGH);
      SET_INPUT(MISO_PIN);
-     WRITE(MISO_PIN,1);
-    #else
+    WRITE(MISO_PIN, HIGH);
+    //#else
      OUT_WRITE(SS_PIN, HIGH);
-    #endif
+    //#endif

    OUT_WRITE(MAX6675_SS, HIGH);

  #endif //HEATER_0_USES_MAX6675
```

Then, I guess that initialize process of Marlin has some kind of problem...

**Comment by Blue-Marlin** [comment]
Recently we shifted the initialisation of the thermal manager in front of the initialisation of the display/SDcard, because of uncontrolled heaters during the boot screens.

@esenapaj 
Your 'hack' makes perfect sense to me. The SPI interface has to be initialized here.

**Comment by ghost** [comment]
My final result at this time: https://github.com/esenapaj/Marlin/commit/a76393d6a0c3fc844359c424b562765cdafca4e8
In `Marlin_main.cpp`

``` diff
  lcd_init();
  #if ENABLED(SHOW_BOOTSCREEN)
    #if ENABLED(DOGLCD)
      safe_delay(BOOTSCREEN_TIMEOUT);
    #elif ENABLED(ULTRA_LCD)
      bootscreen();
-      lcd_init();
+      #if DISABLED(SDSUPPORT)
+        lcd_init();
+      #endif
    #endif
  #endif
```

This second `lcd_init()` causes freeze when `SDSUPPORT` (and MAX31855) is enabled for some reason.
But when `SDSUPPORT` is disabled, this doesn't causes freeze, moreover if this second `lcd_init()` is removed, displaying of LCD is broken.

In `tmperature.cpp`

``` diff
  #if ENABLED(HEATER_0_USES_MAX6675)

-    #if DISABLED(SDSUPPORT)
-      OUT_WRITE(SCK_PIN, LOW);
-      OUT_WRITE(MOSI_PIN, HIGH);
-      SET_INPUT(MISO_PIN);
-      WRITE(MISO_PIN,1);
-    #else
-      OUT_WRITE(SS_PIN, HIGH);
-    #endif
+    OUT_WRITE(SCK_PIN, LOW);
+    OUT_WRITE(MOSI_PIN, HIGH);
+    SET_INPUT(MISO_PIN);
+    WRITE(MISO_PIN, HIGH);
+    OUT_WRITE(SS_PIN, HIGH);

    OUT_WRITE(MAX6675_SS, HIGH);

  #endif //HEATER_0_USES_MAX6675
```

By test, these all the writing process are needed whether `SDSUPPORT` is enabled or not.
If one of those lack, MINTEMP Err or freeze occures.
I wonder that why `OUT_WRITE(SS_PIN, HIGH)` is also needed when `SDSUPPORT` is disabled.

This result is still bit unstable (broken displaying or slow bootup or MAXTEMP Err occures once in a while when bootup),
but in that case, remove a SDCard then reset or completely power off / on the machine, it restore stability.

**Comment by ghost** [comment]
@derletztename 
I prepared a test branch that it's based on newest RCBugFix.
Please try it if you'd like.
~~https://github.com/esenapaj/Marlin/tree/succeed~~

**Comment by derletztename** [comment]
So great, i absolutely will test this. Thanks

**Comment by derletztename** [comment]
Your code is working on my arduino, it doesn't freeze anymore, but I got an Mintemp error and Marlin stops working immediately. I checked the wiring again and everything is fine. Any Ideas about that?

**Comment by ghost** [comment]
I guess that maybe your wiring of K thermocouple is inverted.
[K thermocouple has polarity](http://digital.ni.com/public.nsf/allkb/29FC44A9C662D80186256B02008387CF).
When start heating with inversely wired K thermocouple, 
measured temperature isn't increased but decreased, finally it go below the mintemp.

And this is information: my wiring
![img_20161015_203011_](https://cloud.githubusercontent.com/assets/15308844/19410003/fab86e42-931b-11e6-8799-9957e8b2504a.jpg)

**Comment by derletztename** [comment]
Actually it was at first, but after changing it, the error stayed the same. Also it's popping up just after booting Marlin via Pronterface. So I have no chance to see where the temperature would go to (up or down). It's just killing the Process with an error. Maybe I could upload a screenshot...

**Comment by derletztename** [comment]
![wp_20161015_14_21_33_rich](https://cloud.githubusercontent.com/assets/7712645/19410100/b6269548-92e3-11e6-8df4-9af756a8ee6c.jpg)

**Comment by ghost** [comment]
This is mine.
![default](https://cloud.githubusercontent.com/assets/15308844/19410217/1f5d6630-9321-11e6-893f-60d290ce0134.png)

I'd like to see your `Configuration.h`, `Configuration_adv.h`, and `pins_RAMPS.h`.
If you rename those files to `Configuration.txt`, `Configuration_adv.txt`, and `pins_RAMPS.txt`, and drag & drop to Comment window in this thread, those files are uploaded.
I can test your setting on my machine, because we has same hardware (RAMPS and MAX31855 by Adafruit).

**Comment by ghost** [comment]
By the way, why your compiled date is Aug 17 2016...?

**Comment by derletztename** [comment]
this is wierd. I just downloaded your https://github.com/esenapaj/Marlin/tree/succeed and just changed their in the config.h the setting for thermistor to -3

**Comment by derletztename** [comment]
I use it on a PC that never has been connected to the web, it has a wrong date in the system, I never noticed it^^

**Comment by derletztename** [comment]
Btw why is there a 1 for thermistor and not a -3 as this is your system? Maybe other things are wrong too then?

**Comment by derletztename** [comment]
Where are you from? Can I get your hardware setup with software and just change everything at my place?^^ These little software issues really killing me :(

**Comment by ghost** [comment]
Unfortunately, I live in 9000 kilometers to the east of your country...

By the way, I found two new cause of Mintemp error on bootup.
Those are "poor contacting" or "broken port".
When 5V (Vin) or SCK (CLK) is contacted poorly or disconnected, Mintemp error occurs.
On the other hand when GND or MISO (DO) or MAX6675_SS (CS) is contacted poorly or disconnected, Maxtemp error occurs.

**Comment by derletztename** [comment]
Too bad you live that far away^^
So you mean contacted physically poor? I just use plug-in connectors for the experimental setup, you recommend soldering than?

**Comment by derletztename** [comment]
To get that straight, the version you uploaded to github has thermistor settings for "**heater 0**" set to "**1**" instead of "**-3**", as we both use it. Why is this? Maybe some other settings differ from the version you are actually using? Or maybe you forget to change to thermocouple usage and just measure with an thermistor you never disconnected?

**Comment by ghost** [comment]
> So you mean contacted physically poor?

Yes it is, but you said that you use plug-in connectors, then maybe possibility of poor contacting is none...
...
Would you take a picture of your wiring of RAMPS and MAX31855, if it's possible?
I'm not sure that it's useful for solving the problem, but I can't conceive cause no longer.

> the version you uploaded to github has thermistor settings for heater 0 set to 1instead of -3 as we both use it. Why is this?

It's a "plane" (no modified from official Marlin) setting. It's premised that you alter configs by yourself.
[This is my setting for test](https://github.com/esenapaj/Marlin/commit/7604c6c1d904b47f21d83254888e6e59ec85a4ee), only change the `TEMP_SENSOR_0` from 1 to -3.

> you forget to change to thermocouple usage and just measure with an thermistor you never disconnected?

No, it isn't.
Please look at my picture of RAMPS on above, any thermistor isn't connected.

**Comment by Grogyan** [comment]
I can't tell from the picture if you connected to the 5v on the Max31855
board.
Please also ensure that you have the CS pin connected to pin 66 on RAMPS

**Comment by derletztename** [comment]
![wp_20161015_21_22_34_rich](https://cloud.githubusercontent.com/assets/7712645/19412804/c46a5a1a-931d-11e6-8792-850acb309273.jpg)
![wp_20161015_21_22_23_rich](https://cloud.githubusercontent.com/assets/7712645/19412805/c499361e-931d-11e6-80ed-d5376775b0d1.jpg)

**Comment by ghost** [comment]
@derletztename 
Your wiring is wrong.
**SCK** (D52, purple wire) should be connected to **CLK** of MAX31855.
**MAX6675_SS** (D66, gray) should be connected to **CS** of MAX31855.

But You're connecting **SCK** to **CS**, and **MAX6675_SS** to **CLK**.
That wrong wiring causes Mintemp error.
I've confirmed it just now.

Edit:
...Um?
It looks like that gray wire is connected to black wire, purple wire is connected to red wire, and yellow wire is connected to orange wire.
But it looks like that those black, red, and orange wires aren't connected to MAX31855... ?

**Comment by derletztename** [comment]
Dude, I love you. Sometimes you're not able to "see the forest for the trees", when you look on something too long

**Comment by derletztename** [comment]
Hm, I understand that it actually was wrong, but after changing the wiring it's still the same error. How could it still be the same errror?

**Comment by derletztename** [comment]
Oh I hadn't seen your edit until now. Yes I had to extend all those wires. But I doublechecked the extension and I am very sure they are connected right

**Comment by ghost** [comment]
Um... possibility of problem... possibly signal trouble by long digital line (wire between RAMPS and MAX31855)?
Because from your pictures, I feel that digital line of your system is long. At least, it seems that it's not short.
But SPI is [for short distance communication](https://en.wikipedia.org/wiki/Serial_Peripheral_Interface_Bus). It usually works in under 1m range.
If you need long distance between thermocouple and RAMPS, I recommend that using Compensation Lead Wires for K Thermocouple, instead of  long SPI line. 
If you have short jumper wire, would you test with it?

**Comment by derletztename** [comment]
I did not know this. I will try with short cables now. Though the cables are shorter than 1 m

**Comment by derletztename** [comment]
It's working now. Man I really need someone who understand this stuff properly. Thank you soo much. I tested so many combinations, but never used shorter cables...

**Comment by ghost** [comment]
It's good for you, really good for you (and me)...
It's better that SPI cable is shorter as far as possible.

Thus, I'll submit the PR, by using above test branch.

**Comment by Grogyan** [comment]
Glad to hear you got it working now.
Yeah,  SPI is meant for short communication distances. Depending on the
device you are communicating to,  this is up to about 20cm.
It still worries me that the unshielded cable for SD card handling on many
3D printers are so long.
There are other protocols that allow for error correction such as
Rs232/RS485,  which require dedicated devices to translate the protocol on
the peripheral.

**Comment by derletztename** [comment]
![wp_20161016_15_11_48_rich](https://cloud.githubusercontent.com/assets/7712645/19417674/30ef7e00-93b3-11e6-9e71-c825be2405c1.jpg)

**Comment by derletztename** [comment]
I found a new errror when I reached 1000°C. Maybe the little breakout board was to near to the source? How warm can it get? It maybe reached about 30°C, I couldn't measure it properly

**Comment by derletztename** [comment]
esenapaj my saviour please help me again

**Comment by derletztename** [comment]
I set Maxtemp to 1650°C, but the real error was this MAX31855 Short to VCC thing at first. After the printer halt I disconnected and connected again and got exactly 1000°C. Maybe it's something about this number?

**Comment by Blue-Marlin** [comment]
https://github.com/MarlinFirmware/Marlin/blob/RCBugFix/Marlin/temperature.cpp#L672
Thats the hardware linit for a MAX6675.

**Comment by ghost** [comment]
I prepared a test branch again.
The altering in this branch extend a measuring range of MAX31855 from 1047℃ to 1799℃, maybe.
https://github.com/esenapaj/Marlin/tree/max31855-test-again

But I think that "Short to VCC" error is another problem from MAXTEMP Error.
From [datasheet](https://cdn-shop.adafruit.com/datasheets/MAX31855.pdf), "Short to VCC" (SCV Fault) causes when the thermocouple is short-circuited to VCC (power line).
Basically, it means that something of problem of hardware (MAX31855 chip or board, wire of thermocouple, etc) occurred.

I can't test about this problem at this time, because I haven't high-power burner or furnace that it realize over 1000℃...

edit: ~~um? compilation errors are occurring. wait a moment.~~ fixed.

**Comment by derletztename** [comment]
Today I had the chance to test again and I surpassed 1000 °C :)
But it failed again at something like 1400°C :( With the exact same error
Maybe we can push the boundarys even a little higher? Steel needs about 1500°C to melt and therefore you need something like 1550 to 1600 for printing I guess...

**Comment by ghost** [comment]
Are you still using stocked (non-modified) Adafruit board and K thermocouple?
If it's true, probably 1400℃ is limitation of MAX31855KASA and K thermocouple.
From [datasheet](https://cdn-shop.adafruit.com/datasheets/MAX31855.pdf),

```
K Alumel Chromel -270 to +1372
S Platinum Platinum/Rhodium +50 to +1768
R Platinum Platinum/Rhodium -50 to +1768
```

.

Or, now are you using a modified board that it was resoldered to MAX31855SASA, and S thermocouple?

**Comment by derletztename** [comment]
As the original board worked I also tested the modified one and it worked also. So right now I am using the modified one without fully understanding the consequences of modifying it. I really need a team and someone that understands the electronic...
To my understanding when the chip is changed, everything is changed as the rest are only small resistors and capacitors to reduce noise and stuff like that

**Comment by ghost** [comment]
Well, I'm sorry but it's unclear for me,
it means that now you're measuring over 1400℃ with modified board and S thermocouple without problem?

**Comment by derletztename** [comment]
Sorry was writing on my phone while watching the kids. They are always in trouble ;)
Now I am using the modified version with the MAX31855SASA soldered in and the S type thermocouple. Thanks to your help I am measuring temperatures above 1000°C. But I can't get above some point around 1400°C. Maybe it's about 1732°C, I don't know exactly, I was looking at the nozzle at that moment :(

**Comment by ghost** [comment]
I've done a experiment that it uses cheating as follows.
https://github.com/esenapaj/Marlin/commit/3021bb1f5feee58e6bc136ccff74d1e042ee35d9
https://github.com/esenapaj/Marlin/tree/max31855-test-again-cheating

This cheating return 1799℃ always, and I confirmed that Marlin can handle the over 1400℃ temperature.
![default](https://cloud.githubusercontent.com/assets/15308844/19531524/fa5212b4-9672-11e6-9dca-ad38e63230fe.png)

Therefore, I think that ”1400℃ issue" isn't related to software but hardware.

By the way, you said that you're using a "chinese thermocouple type S", is it this?

[Durable S type thermocouple WRP 100 platinum and rhodium thermocouple probe length 225mm probe head temperature sensor 1300 C-in Temperature Instruments from Home Improvement on Aliexpress.com | Alibaba Group](https://www.aliexpress.com/item/Durable-S-type-thermocouple-WRP-100-platinum-and-rhodium-thermocouple-probe-length-225mm-probe-head-temperature/32702775310.html)

If it's true, the description says that "Measuring range: 0-**1300**℃"...

**Comment by ghost** [comment]
And, Adafruit is selling MAX3185 **"6"** board, it works with any K, J, N, R, S, T, E, or B type thermocouple.
[Adafruit Universal Thermocouple Amplifier MAX31856 Breakout ID: 3263 - $17.50 : Adafruit Industries, Unique & fun DIY electronics and kits](https://www.adafruit.com/product/3263)

...But yes, Marlin doesn't support the MAX31856 yet...

**Comment by derletztename** [comment]
Yeah, I use this thermocouple. I just broke the case/shell apart and use only the thermocouple tip for better measuring performance. The wires are still in another ceramic casing, therefore not short circuited. 
After checking again, the thermocouple maybe just go up to 1400°C. I depends on the materialmix and I should assume they've chosen the cheapest one...

**Comment by derletztename** [comment]
I think I need to invest into a better one now. They are just so incredibly expensive in germany.
Also the Board you postet is very great. I really love to use this one. Now I only need someone to implement it into Marlin. Can't be that difficult actually as Adafruit is giving a library and some explanations with it, but still I have no clue of coding :(

**Comment by derletztename** [comment]
I ordered a new and way more professional thermocouple type S, it wasn't as expensive as I thought and there is no issue using it permanently at around 1600°C. I also bought one of those boards from adafruit. You maybe have some sparetime to help me with the coding esenapaj? I promise to print you many metal hearts as soon as the machine is working ;)

**Comment by ghost** [comment]
Well, you say that you purchased MAX31856 board already?

**Comment by derletztename** [comment]
Yes, why?

**Comment by ghost** [comment]
I've been surprised your enthusiasm, because Marlin doesn't support MAX31856 yet.
But on the other hand, the enthusiasm is agreeable for me...

So I start trying to implement the MAX31856, but it seems that it isn't easy for me.
Maybe it won't go smoothly...

**Comment by derletztename** [comment]
If there is a will, there will be a solution ;)
If I can help, I'd love too. Thanks for supporting the project, thats so great. Is there a way to communicate more directly? Maybe have "our" own development area? So we do not spam the complete marlin git and just release results afterwards?

**Comment by derletztename** [comment]
Btw. the company, assembling the thermocouple. just forecasted the 27.10.16 as delivery date :)

**Comment by ghost** [comment]
Good news for you (and me?), @thinkyhead will implement supporting for MAX31856 to Marlin at some point pretty soon.

[Request] Support for MAX31856 · Issue #5046 · MarlinFirmware/Marlin

> At some point pretty soon I will put in the time to implement this.

**Comment by derletztename** [comment]
So do we close this issue, as there is no further need of the MAX31855 for measuring temperatures above 1400°C ?

**Comment by ghost** [comment]
About measuring over 1023℃ temperature by MAX31855, I submit a patch today (Extend measuring range of thermocouple for MAX31855 #5071).
This patch extend for measuring range of MAX31855 up from 1023℃ to 1799℃(limit of MAX31855 chip - 1℃).
Limit of Adafruit MAX31855 board is 1350℃ because it's K thermocouple version (MAX31855KASA),
but [another MAX31855 board](http://www.ebay.com/itm/MAX31855-K-T-R-J-Type-Thermocouple-Module-arduino-compatible-MAX6675-/111702003419) that it can measure 1770℃ exist on market, it's R thermocouple version (MAX31855RASA).

But, MAX31855RASA board and R thermocouple isn't relaterd to you currently, because you'll use a MAX31856 board and S thermocouple.
Therefore, this thread can be close if you'd like to do it.

**Comment by derletztename** [comment]
Thanks esenapaj :)

**Comment by derletztename** [comment]
Hey Guys
It has been quite a while. Sorry for that.
So in the meantime, I got a new professional thermocouple, the first one had an issue and the manufacturer made me a new one. This seems to work find now :)
Though, still with the MAX31856 I am measuring very slow raising numbers and too low temperatures. Should be in the hundreds, but is only about 45°C.
And with the modified MAX31855 board, I again face the 1000°C max limit, which I fail to solve again. I am not sure which version was the working one, but the RCBugFix won't help me with the 1000°C limit. Though I remember we were beyond this, the last barrier we face was 1400°C and this was an issue with the thermocouple(s). Could you maybe help me again with this, esenapaj?
